### DISCUSSION TITLE
servicediscovery compatability issue with localstack

### DISCUSSION THREAD

**Discussion by codefromthecrypt** [aws/aws-sdk-go-v2#2557] [Q&A]
### Acknowledgements

- [X] I have searched (https://github.com/aws/aws-sdk/issues?q=is%3Aissue) for past instances of this issue
- [X] I have verified all of my SDK modules are up-to-date (you can perform a bulk update with `go get -u github.com/aws/aws-sdk-go-v2/...`)

### Describe the bug

DiscoverInstances prefixes "data-" to the hostname, rendering localhost (specifically localstack) tests useless, as the hostname "data-localhost" doesn't exist.

### Expected Behavior

I would like to use all servicediscovery functions. I'm not sure why others work and DiscoverInstances specifically does this

### Current Behavior

I end up with an error like this:

```
2024-03-14T12:39:51.350703Z	error	unable to refresh Cloud Map cache due to error, using existing cache: error retrieving instance list from Cloud Map for "1" in "testns": operation error ServiceDiscovery: DiscoverInstances, https response error StatusCode: 0, RequestID: , request send failed, Post "http://data-localhost:32957/": dial tcp: lookup data-localhost: no such host
```

### Reproduction Steps

localstack requires an authenticated docker image (localstack-pro) to run, so probably reproduction won't help as much, but the command that fails looks like this:

```go
	instOutput, err := client.DiscoverInstances(ctx, &servicediscovery.DiscoverInstancesInput{ServiceName: svc.Name, NamespaceName: ns.Name})
	if err != nil {
		return nil, errors.Wrapf(err, "error retrieving instance list from Cloud Map for %q in %q", *svc.Name, *ns.Name)
	}
```

### Possible Solution

maybe there is a way to avoid this part of the stack. I don't mind a workaround for testing.

### Additional Information/Context

where the prefix happens: https://github.com/aws/aws-sdk-go-v2/blob/main/service/servicediscovery/api_op_DiscoverInstances.go#L193

### AWS Go SDK V2 Module Versions Used

```
	github.com/aws/aws-sdk-go-v2 v1.25.3
	github.com/aws/aws-sdk-go-v2/config v1.27.7
	github.com/aws/aws-sdk-go-v2/credentials v1.17.7
	github.com/aws/aws-sdk-go-v2/service/servicediscovery v1.29.2
```

### Compiler and Version used

go version go1.22.1 darwin/arm64

### Operating System and version

macOS 14.4

**Comment by lucix-aws** [comment]
The AWS SDKs are designed for and are only guaranteed to work against the public AWS services. The `servicediscovery` service apparently uses `data-` as a host prefix for certain operations.

If you're using an alternative implementation of the service that doesn't work this way, you can technically remove the `EndpointHostPrefix` middleware from the `Finalize` step, though I can't promise that won't cause other issues:

```go
svc := servicediscovery.NewFromConfig(cfg, func (o *servicediscovery.Options) {
    o.APIOptions = append(o.APIOptions, func (stack *middleware.Stack) error {
        // this technically returns an error if the middleware doesn't exist, and on some servicediscovery
        // ops that IS the case, but we don't care
        stack.Finalize.Remove("EndpointHostPrefix")
        return nil
    })
})
```

**Reply by codefromthecrypt** [reply]
Thanks for the workaround.

An aside which may or may not change your opinion on handling this. I found out about this approach through work of an AWS employee. It could make sense to make testing possible as attempted here in java by @hariohmprasath https://github.com/awspring/spring-cloud-aws/pull/506 Basically, go shouldn't have a harder time than java imho, but up to you

**Reply by codefromthecrypt** [reply]
also thanks so much for how fast you got me going. I had the luxury of waking up and able to progress. champion stuff!
